content/blog/posts/2022-01-13-react-objects-and-state.md

```diff
@@ -0,0 +1,171 @@
+        onChange={(e) => handleInputChange(e, 'name')}
+      />
+    </label>
+    <label>
+      Age:
+      <input
+        value={person.age}
+        onChange={(e) => handleInputChange(e, ‘age’)}
+      />
+    </label>
+    <p>
+      {user.name}: {user.age}
+    </p>
+  </>
+)
+```
+
+### Use a Library
+
+If you find yourself using multiple objects or in your state, it may be worth it to install a library like [immer](https://github.com/immerjs/use-immer). Immer provides a proxy object, referred to as a `draft`, so while your code appears similar to a direct mutation of state, a copy is created under the hood when an update when immer is used.
+
+#### Why Won’t Object.assign() Update State?
+
+The `Object.assign()` method takes 2 arguments - a `target` and at least one `source` and returns the `target` after copying the `source`s’ properties. [Object.assign()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) won’t trigger a re-render unless the `target` provided is a brand new object. When the `target` is the original object, it still has the same identity, similar to updating a property directly. At this point, the syntax gets clunkier too, so it is likely easier to read without this additional method.
+
+```
+// This would not update state
+setUser(Object.assign(user, userName))
+
+// You’d still need a reference to a new object for this to update
+setUser(Object.assign({}, user, userName))
+```
+
+## So When Should an Object Be Used in State?
+
+Following basic programming principles, break down state into “atoms” and use simpler values (primitives) when possible.
+
+When values in a component are unrelated, it’s also helpful to separate instances of useState to visually signal your intentions. A user altering their name or age probably will not affect one another and therefore it makes more sense to be stored in 2 separate instances of `useState`. This helps simplify the component logic so that it is easier to read the overall logic as well.
+
+```
+const [name, setName] = useState(‘’)
+const [age, setAge] = useState(0)
+
+return (
+  <>
+    <label>
+      Name:
+      <input
+        value={name}
+        onChange={(e) => setName(e.target.value)}
+      />
+    </label>
+    <label>
+      Age:
+      <input
+        value={age}
+        onChange={(e) => setAge(e.target.value)}
+      />
+    </label>
+    <p>
+      {name}: {age}
+    </p>
+  </>
+);
+```
+
+When you have[ three or more values](https://thoughtspile.github.io/2021/10/11/usestate-object-vs-multiple/) you need to keep track of, your app can yield better runtime performance with a single object state than numerous instances useState. This approach may be better for something like a form with many inputs. When you have something like a controlled form component with many inputs that will all get saved or updated to one API, it may make visual sense to store these data in one object since they will reuse the same logic.
```

Comment:
```suggestion
When you have [three or more values](https://thoughtspile.github.io/2021/10/11/usestate-object-vs-multiple/) you need to keep track of, your app can yield better runtime performance with a single object state than numerous instances useState. This approach may be better for something like a form with many inputs. When you have something like a controlled form component with many inputs that will all get saved or updated to one API, it may make visual sense to store these data in one object since they will reuse the same logic.
```